provider-example/src/hpke.rs

```diff
@@ -0,0 +1,148 @@
+    fn supports_aead(&self, aead: HpkeAead) -> bool {
+        match hpke_rs_crypto::types::AeadAlgorithm::try_from(aead.get_u16()).ok() {
+            Some(hpke_rs_aead) => HpkeRustCrypto::supports_aead(hpke_rs_aead).is_ok(),
+            None => false,
+        }
+    }
+}
+
+struct HpkeRsHpke(hpke_rs::Hpke<HpkeRustCrypto>);
+
+impl Debug for HpkeRsHpke {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("HpkeRsHpke").finish()
+    }
+}
+
+impl Hpke for HpkeRsHpke {
+    fn setup_sender(
+        &self,
+        pk_r: &HpkePublicKey,
+        info: &[u8],
+        psk: Option<&HpkePreSharedKey>,
+        psk_id: Option<&[u8]>,
+        sk_s: Option<&HpkePrivateKey>,
+    ) -> Result<(EncapsulatedSecret, Box<dyn HpkeSender + '_>), Error> {
+        let pk_r = hpke_rs::HpkePublicKey::new(pk_r.0.clone());
+        let psk = psk.map(|psk| psk.0.as_slice());
+        let sk_s = sk_s.map(|sk_s| hpke_rs::HpkePrivateKey::new(sk_s.0.clone()));
+        let (enc, context) = self
+            .0
+            .setup_sender(&pk_r, info, psk, psk_id, sk_s.as_ref())
+            .map_err(general_err)?;
+        Ok((
+            EncapsulatedSecret(enc.to_vec()),
+            Box::new(HpkeRsSender { context }),
+        ))
+    }
+
+    fn setup_receiver(
+        &self,
+        enc: &EncapsulatedSecret,
+        sk_r: &HpkePrivateKey,
+        info: &[u8],
+        psk: Option<&HpkePreSharedKey>,
+        psk_id: Option<&[u8]>,
+        pk_s: Option<&HpkePublicKey>,
+    ) -> Result<Box<dyn HpkeReceiver + '_>, Error> {
+        let sk_r = hpke_rs::HpkePrivateKey::new(sk_r.0.clone());
+        let psk = psk.map(|psk| psk.0.as_slice());
+        let pk_s = pk_s.map(|pk_s| hpke_rs::HpkePublicKey::new(pk_s.0.clone()));
+        let context = self
+            .0
+            .setup_receiver(enc.0.as_slice(), &sk_r, info, psk, psk_id, pk_s.as_ref())
+            .map_err(general_err)?;
+        Ok(Box::new(HpkeRsReceiver { context }))
+    }
+}
+
+#[derive(Debug)]
+struct HpkeRsSender {
+    context: hpke_rs::Context<HpkeRustCrypto>,
+}
+
+impl HpkeSender for HpkeRsSender {
+    fn seal(&mut self, aad: &[u8], plaintext: &[u8]) -> Result<Vec<u8>, Error> {
+        self.context
+            .seal(aad, plaintext)
+            .map_err(general_err)
+    }
+}
+
+#[derive(Debug)]
+struct HpkeRsReceiver {
+    context: hpke_rs::Context<HpkeRustCrypto>,
+}
+
+impl HpkeReceiver for HpkeRsReceiver {
+    fn open(&mut self, aad: &[u8], ciphertext: &[u8]) -> Result<Vec<u8>, Error> {
+        self.context
+            .open(aad, ciphertext)
+            .map_err(general_err)
+    }
+}
+
+fn general_err(err: impl Debug) -> Error {
+    // TODO(@cpu): It would be better to use `Error::Other` here, but it requires `Error` to be
+    //   implemented, and `HpkeError` doesn't.
+    // TODO(@cpu): It would be nicer to user `err.to_string()` here instead of `format!` but
+    //   it requires `Display` to be implemented, and HpkeError doesn't.
```

Comment:
I think this is stale and leftover from when I was using an older version of hpke-rs. I'll adjust.